io/download_test.ts

```diff
@@ -0,0 +1,192 @@
+import { downloadFile, downloadFileToTempDir } from './download.ts';
+import { crypto, iterateReader, path } from '../deps.ts';
+import {
+  assertEquals,
+  assertInstanceOf,
+  assertNotEquals,
+  createStaticTestHttpServer,
+  createTestHttpServer,
+  deterministicByteArray,
+  fail,
+  fileAttachmentHeaders,
+} from '../testing_deps.ts';
+
+const TEST_FILE_CONTENTS = 'testing 1 2 3...';
+
+Deno.test('downloads text file', async () => {
+  const testServer = createStaticTestHttpServer(TEST_FILE_CONTENTS);
+
+  const downloadedFile = await downloadFileToTempDir(testServer.urlOrigin);
+
+  const fileContents = await Deno.readTextFile(downloadedFile);
+  assertEquals(fileContents, TEST_FILE_CONTENTS);
+
+  await testServer.close();
+  await Deno.remove(downloadedFile);
+});
+
+Deno.test('downloads binary file', async () => {
+  // 32MB feels sufficiently large to test some buffering, but not large enough to be obnoxious?
+  const binaryData = deterministicByteArray(32 * 1024 * 1024);
+  const expectedHash = new Uint8Array(await crypto.subtle.digest('SHA-256', binaryData));
+  const testServer = createStaticTestHttpServer(binaryData);
+
+  const downloadedFile = await downloadFileToTempDir(testServer.urlOrigin);
+
+  const file = await Deno.open(downloadedFile, { read: true });
+  try {
+    const hash = new Uint8Array(await crypto.subtle.digest('SHA-256', iterateReader(file)));
+    assertEquals(hash, expectedHash);
+  } finally {
+    await testServer.close();
+    file.close();
+    await Deno.remove(downloadedFile);
+  }
+});
+
+Deno.test('still works with no sensible default filename', async () => {
+  const testServer = createStaticTestHttpServer(TEST_FILE_CONTENTS);
+
+  const downloadedFile = await downloadFile({
+    url: testServer.urlOrigin,
+    downloadLocation: { specialDir: 'temp' },
+  });
+
+  try {
+    const baseFilename = path.basename(downloadedFile);
+    assertNotEquals(baseFilename, '');
```

Comment:
Done, but I didn't do this because I try to have tests only test one thing. The reasoning for this is that, if we break the download integrity and this test fails, it might look like the strategy for choosing filenames is broken too.